swan-lake/development-tutorials/test-document-the-code/ballerina-profiler.md

```diff
@@ -0,0 +1,365 @@
+            return false;
+        }
+    }
+    return true;
+}
+
+isolated function createRandomIntArray(int size) returns int[]|error {
+    int[] array = [];
+    foreach int i in 0 ..< size {
+        array.push(check random:createIntInRange(0, int:MAX_VALUE));
+    }
+    return array;
+}
+```
+
+Run and profile the Ballerina package using the `bal profile` CLI command.
+
+```
+$ bal profile
+```
+
+This will give the following output.
+
+```
+Compiling source
+        profiler_demo/sort:0.1.0
+
+Generating executable
+        target/bin/sort.jar
+
+================================================================================
+Ballerina Profiler: Profiling...
+================================================================================
+Note: This is an experimental feature, which supports only a limited set of functionality.
+[1/6] Initializing...
+[2/6] Copying executable...
+[3/6] Performing analysis...
+[4/6] Instrumenting functions...
+ ○ Instrumented module count: 31
+ ○ Instrumented function count: 1016
+[5/6] Running executable...
+Is the array sorted? true
+[6/6] Generating output...
+ ○ Execution time: 3 seconds 
+ ○ Output: target/bin/ProfilerOutput.html
+-------------------------------------------------------------------------------
+```
+
+Open the `target/bin/ProfilerOutput.html` file using a web browser window to examine the Ballerina Profiler output to find the slow-running functions and performance bottlenecks, which can then be addressed to improve the program’s overall performance.
+
+## Profiler output and available details
+
+![Profiler output](/learn/images/profiler-output.png "Profiler output")
+
+The flame graph generated by the Ballerina Profiler contains the following details.
+
+1. Function call stack
+
+2. Time taken to execute each function and its percentage
+
+Flame graph also has the following features.
+
+1. Search a function.
+
+![Search function inside flame graph](/learn/images/search-function-inside-flame-graph.png "Search function inside flame graph")
+
+2. Clear the search.
+
+3. Click on a function call and zoom the view.
+
+![Click function and zoom flame graph](/learn/images/click-function-and-zoom-flame-graph.png "Click function and zoom flame graph")
+
+4. Reset the view.
+
+## Examine the output and find performance bottlenecks
+
+When we observe the flame graph generated by the Ballerina Profile for the above code, we can see the `bubbleSort` function has taken the most time to execute.
+
+![Bubble sort function time](/learn/images/bubble-sort-function-time.png "Bubble sort function time")
+
+Let's try to optimize the sorting function to increase the performance of our Ballerina application.
+
+## Optimize the Ballerina code
+
+In this example, we will use another sorting algorithm to reduce the time taken to execute our application. Let’s replace the `main.bal` file with the code below which uses the merge sort algorithm instead of the bubble sort algorithm for sorting.
```

Comment:
```suggestion
In this example, you will use another sorting algorithm to reduce the time taken to execute your application. Replace the `main.bal` file with the code below, which uses the `merge sort` algorithm instead of the `bubble sort` algorithm for sorting.
```

Reply:
Since bubble sort and merge sort is just a name not a code snippet here will not add **`** as done here https://en.wikipedia.org/wiki/Merge_sort.